internal/proxy/clientgo_cached_proxy_resolver.go

```diff
@@ -155,6 +155,17 @@ func (p *clientgoCachedProxyResolver) ObjectExists(obj client.Object) (bool, err
 	return exists, err
 }
 
+func (p *clientgoCachedProxyResolver) IsReady() bool {
+	// If the proxy is has no database, it is only ready after a successful sync
+	// Otherwise, it has no configuration loaded
+	if p.dbmode == "off" {
```

Comment:
Is there a possibility that more than one goroutine will be accessing the lastConfigSHA variable? If so, we should synchronize access by wrapping the access code with a mutex.

Reply:
Yes, the sync loop and readiness handler server are separate goroutines, but I'm unsure what issue actually arises beyond the race checker complaining, maybe (it apparently didn't care when CI ran). While proxy sync loop routine can set `lastConfigSHA`, the readiness handler can only read it, so at worst it tries to read as a write is occurring and gets outdated info.

Outdated info is fine, since the readiness check will just run again and see the new value after.

Reply:
This is a read operation against the SHA, per its definition, it indicates a successful POS to kong, which satisfy the needs. 
I don't think any con-current check impact the check, it is not the interested check point.

Reply:
I would propose a clean POST config call to proxy within the IsReady function as following 
https://github.com/Kong/kubernetes-ingress-controller/blob/main/internal/sendconfig/sendconfig.go#L143
and gets readiness from the response.

Reply:
> I'm unsure what issue actually arises beyond the race checker complaining

That's a very good question and I did a bit of online search to find a good answer.

*Benign Data Races: What Could Possibly Go Wrong?*
https://software.intel.com/content/www/us/en/develop/blogs/benign-data-races-what-could-possibly-go-wrong.html

> The bottom line. Just say No to “benign” races. Even if you still think that that particular data race is 100% safe (which I doubt), it’s still formally incorrect, fragile during code maintenance and produces noise under race detection tools.
Use automatic data race detection tools, like ThreadSanitizer or Intel Parallel Inspector. Use them regularly during continuous testing (like it’s done by the Chromium project). Test your systems under realistic workloads with the tools to find more bugs. And, of course, fix the races. That will make your life much easier and will save your team lots of time.

Now that we have identified the data race, fixing it is very easy: just synchronize the access to the shared variable.

Does the above explanation sound convincing?

Reply:
Part of 93387e8b.